pkg/controller/proxyconfig/controller.go

```diff
@@ -0,0 +1,361 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager, status *statusmanager.StatusManager) error {
+	reconciler := newReconciler(mgr, status)
+	if reconciler == nil {
+		return fmt.Errorf("failed to create reconciler")
+	}
+
+	return add(mgr, reconciler)
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, status *statusmanager.StatusManager) reconcile.Reconciler {
+	if err := configv1.Install(mgr.GetScheme()); err != nil {
+		return &ReconcileProxyConfig{}
+	}
+
+	return &ReconcileProxyConfig{client: mgr.GetClient(), scheme: mgr.GetScheme(), status: status}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("proxyconfig-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// We only care about a configmap source with a specific name/namespace,
+	// so filter events before they are provided to the controller event handlers.
+	pred := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return e.MetaOld.GetName() == names.ADDL_TRUST_BUNDLE_CONFIGMAP &&
+				e.MetaOld.GetNamespace() == names.TRUST_BUNDLE_CONFIGMAP_NS
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return e.Meta.GetName() == names.ADDL_TRUST_BUNDLE_CONFIGMAP &&
+				e.Meta.GetNamespace() == names.TRUST_BUNDLE_CONFIGMAP_NS
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return e.Meta.GetName() == names.ADDL_TRUST_BUNDLE_CONFIGMAP &&
+				e.Meta.GetNamespace() == names.TRUST_BUNDLE_CONFIGMAP_NS
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return e.Meta.GetName() == names.ADDL_TRUST_BUNDLE_CONFIGMAP &&
+				e.Meta.GetNamespace() == names.TRUST_BUNDLE_CONFIGMAP_NS
+		},
+	}
+
+	// Watch for changes to the trust bundle configmap.
+	err = c.Watch(&source.Kind{Type: &corev1.ConfigMap{}}, &handler.EnqueueRequestForObject{}, pred)
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to the proxy resource.
+	err = c.Watch(&source.Kind{Type: &configv1.Proxy{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ReconcileProxyConfig reconciles a Proxy object
+type ReconcileProxyConfig struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver.
+	client client.Client
+	scheme *runtime.Scheme
+	status *statusmanager.StatusManager
+}
+
+// Reconcile expects request to refer to a proxy object named "cluster"
+// in the default namespace or to a configmap object named
+// "user-ca-bundle" in namespace "openshift-config-managed", and will
```

Comment:
this comment is wrong...   i'd expect this controller to be watching 3 things:

1) the trust-bundle configmap in openshift-config-managed  (this CM has a fixed name), because you need to stomp it back to the correct value if it gets updated by something else
2) the additional CA bundle configmap in the openshift-config namespace (this configmap name can change, so you need to detect if it changes, by watching the proxy config resource), because if the bundle changes, you need to re-merge the values w/ the system values and publish it to the CM in (1).
3) the proxyconfig resource (so you can determine the name of the configmap in (2) and perform the merge action on it)

Reply:
@bparees see https://github.com/openshift/cluster-network-operator/pull/271#discussion_r313666739